core/src/main/scala/org/apache/spark/sql/execution/oplog/impl/OpLogRdd.scala

```diff
@@ -468,14 +467,20 @@ class OpLogRdd(
             val deltaColIndex = ColumnDelta.deltaColumnIndex(colIndx, 0)
             val deltaEntry1 = regMap.getEntry(k.withColumnIndex(deltaColIndex))
             val delta1 = if (deltaEntry1 ne null) {
-              DiskEntry.Helper.readValueFromDisk(deltaEntry1.asInstanceOf[DiskEntry],
-                phdrCol).asInstanceOf[ColumnFormatValue].getBuffer
+              val buffer = DiskEntry.Helper.readValueFromDisk(deltaEntry1.asInstanceOf[DiskEntry],
+                phdrCol).asInstanceOf[ColumnFormatValue]
+                  .getValueRetain(FetchRequest.DECOMPRESS).getBuffer
+              if(buffer.isDirect) directBuffers += buffer
+              buffer
             } else null
 
             val deltaEntry2 = regMap.getEntry(k.withColumnIndex(deltaColIndex - 1))
             val delta2 = if (deltaEntry2 ne null) {
-              DiskEntry.Helper.readValueFromDisk(deltaEntry2.asInstanceOf[DiskEntry],
-                phdrCol).asInstanceOf[ColumnFormatValue].getBuffer
+              val buffer = DiskEntry.Helper.readValueFromDisk(deltaEntry2.asInstanceOf[DiskEntry],
+                phdrCol).asInstanceOf[ColumnFormatValue]
+                  .getValueRetain(FetchRequest.DECOMPRESS).getBuffer
+              if(buffer.isDirect) directBuffers += buffer
```

Comment:
Just check if directBuffers size is not too large. Otherwise, we can just return the buffer and keep it alive till the row is decoded.

Reply:
Thanks. Will check.